src/canvas/renderCanvas.js

```diff
@@ -99,206 +96,290 @@ function renderReactMarksAndMeasures(
   canvasHeight,
   canvasStartY,
 ) {
-  let priorityMinY = canvasStartY;
-
-  REACT_PRIORITIES.forEach((priority, priorityIndex) => {
-    const currentPriority = data[priority];
-
-    let baseY = priorityMinY + REACT_GUTTER_SIZE;
-
-    if (currentPriority.events.length > 0) {
-      currentPriority.events.forEach(event => {
-        const showHoverHighlight = hoveredEvent && hoveredEvent.event === event;
-        renderSingleReactMarkOrMeasure({
-          baseY,
-          canvasWidth,
+  // TODO: Compute lanes to render from data? Or just use getLaneHeight to skip lanes
+  const lanesToRender: ReactLane[] = Array.from(
+    Array(REACT_TOTAL_NUM_LANES).keys(),
+  );
+
+  let laneMinY = canvasStartY;
+
+  lanesToRender.forEach(lane => {
+    const baseY = laneMinY + REACT_GUTTER_SIZE;
+
+    // TODO: Draw a separate event gutter. Split into a renderReactMarks function?
+    // if (data.events.length > 0) {
+    //   data.events.forEach(event => {
+    //     const showHoverHighlight = hoveredEvent && hoveredEvent.event === event;
+    //     renderSingleReactMarkOrMeasure({
+    //       baseY,
+    //       canvasWidth,
+    //       context,
+    //       eventOrMeasure: event,
+    //       showGroupHighlight: false,
+    //       showHoverHighlight,
+    //       state,
+    //     });
+    //   });
+
+    //   // Draw the hovered and/or selected items on top so they stand out.
+    //   // This is helpful if there are multiple (overlapping) items close to each other.
+    //   // if (hoveredEvent !== null && hoveredEvent.event !== null) {
+    //   //   renderSingleReactMarkOrMeasure({
+    //   //     baseY,
+    //   //     canvasWidth,
+    //   //     context,
+    //   //     eventOrMeasure: hoveredEvent.event,
+    //   //     showGroupHighlight: false,
+    //   //     showHoverHighlight: true,
+    //   //     state,
+    //   //   });
+    //   // }
+
+    //   baseY += REACT_EVENT_SIZE + REACT_GUTTER_SIZE;
+    // }
+
+    data.measures
+      // TODO: Optimization: precompute this so that we don't filter this array |lanesToRender| times
+      .filter(measure => measure.lanes.includes(lane))
+      .forEach(measure => {
+        const showHoverHighlight =
+          hoveredEvent && hoveredEvent.measure === measure;
+        const showGroupHighlight =
+          hoveredEvent &&
+          hoveredEvent.measure !== null &&
```

Comment:
Next line will throw if `hoveredEvent.measure === undefined`. Would probably just do a falsy check, or if there's a lot of this type of deep property munging ahead, use [idx](https://github.com/facebookincubator/idx).

Reply:
`idx` is interesting. Just too bad JS hasn't gotten the optional chaining operator yet.

Flow should already help to ensure that `hoveredEvent.measure` is `null | ReactMeasure`, but yeah a falsy check won't hurt. I'll change it